README.md
<p align="center"><a href="https://github.com/yelloxing/Web-Studio-Code" target="_blank" rel="noopener noreferrer">
<img width="400" src="https://yelloxing.github.io/Web-Studio-Code/logo.png" alt="Web Studio Code"></a></p>

# 🎉 Web Studio Code - An Editor Used on the Browser Side.

<p align="center">
<a href="https://yelloxing.github.io/npm-downloads/?interval=7&packages=wscode"><img src="https://img.shields.io/npm/dm/wscode.svg" alt="Downloads"></a>
<a href="https://packagephobia.now.sh/result?p=wscode"><img src="https://packagephobia.now.sh/badge?p=wscode" alt="install size"></a>
<a href="https://www.jsdelivr.com/package/npm/wscode"><img src="https://data.jsdelivr.com/v1/package/npm/wscode/badge" alt="CDN"></a>
<a href="https://www.npmjs.com/package/wscode"><img src="https://img.shields.io/npm/v/wscode.svg" alt="Version"></a>
<a href="https://github.com/yelloxing/Web-Studio-Code/blob/master/LICENSE"><img src="https://img.shields.io/npm/l/wscode.svg" alt="License"></a>
</p>

<p align="center"><a href="https://yelloxing.github.io/Web-Studio-Code/Web-Studio-Code.html" target="_blank" rel="noopener noreferrer">
<img width="500" src="https://yelloxing.github.io/Web-Studio-Code/snipping.png" alt="Web Studio Code"></a></p>

> 温馨提示：使用过程中，你可以查看 [版本历史](./CHANGELOG) 来了解是否需要升级！

> 兼容Chrome、Safari、Edge、Firefox、Opera和IE(9+)等常见浏览器！

## 如何引入

我们推荐你使用npm的方式安装和使用：

```bash
npm install --save wscode
```

当然，你也可以通过CDN的方式引入：

```html
<script src="https://cdn.jsdelivr.net/npm/wscode"></script>
```

## 如何使用

- 特别注意：当前最后一个可用版本（非beta和alpha版本）请查看master分支的说明！

```js
import WSCode from 'wscode';

var wscode = new WSCode({

    // 编辑器挂载点(必选)
    el: document.getElementById('wscode'),

    // 初始化文本（可选）
    content: "初始化文本内容",

    // 编辑器字体（可选，默认"新宋体"）
    "font-family": string,

    // 编辑器字重（可选，默认600）
    "font-weight": number,

    // 着色方法（可选，默认不特殊着色）
    shader: function(textString){
        return [
            [{
                content:"内容",
                color:"文字颜色"
            },
            ...],
            ...
        ];
    },

    // 格式化方法（可选）
    format: function(textString, tabNumber){
        return "格式化后的文本";
    },

    // 辅助输入（可选）
    input: function(dom, options, contentArray){
        /*
        1.其中dom和contentArray分别表示辅助的悬浮结点和内容数组;
        2.options的一些重要的辅助信息，是一个json，包括以下内容：
            {
                leftNum:光标前面有多少个字符
                lineNum:当前行之前有多少行
                x:光标left坐标
                y:光标top坐标
                lineHeight:一行文本的高
            }
        */

       // 返回的是键盘操作,可以有任意多个(可选)
       return {
            // keyString可以取:
            //    ”up“:按下键盘向上键
            //    ”down“:按下键盘向下键
            //    等
            // 具体的查看这里： https://yelloxing.github.io/core.js/tools-api/index.html#keyString
            "keyString":function(){

                // 最后返回true或false,默认false表示阻止默认行为（或原有行为）
                return boolean;
            },
            ...
       };
    },

    // 设置颜色（可选）
    color: {
        background: "#d6d6e4", /*编辑器背景*/
        text : "#000000", /*文本颜色*/
        number: "#888484", /*行号颜色*/
        edit: "#eaeaf1", /*编辑行背景色*/
        cursor: "#ff0000", /*光标颜色*/
        select: "#6c6cf1", /*选择背景*/
    },

    // 设置一个tab表示多少个空格（可选，默认4）
    tabSpace: number

});
```

返回的wscode里面挂载着后续可控方法：

- 格式化代码

```js
wscode.format();
```

- 获取当前编辑器代码

```js
wscode.valueOf();
```

- 编辑器管理的文本发生改变后会主动触发callback方法

```js
wscode.updated(callback);
```

- 在当前光标位置输入新的内容

```js
// cursor和length默认都是0，前者表示光标偏移量，后者表示替换个数
wscode.input(content[, cursor, number]);
```

## 开源协议

[MIT](https://github.com/yelloxing/Web-Studio-Code/blob/master/LICENSE)

Copyright (c) 2020 走一步 再走一步


CHANGELOG
No more records!
 -
 -
 -
v1.3.2:
    date:2020-05-13
    changes:
      -  修复由于转义问题导致的<,>和&等特殊字符显示问题
      -  修复由于滚动条导致的点击切行交互错误
v1.3.3:
    date:2020-05-13
    changes:
      -  添加字体font-family可配置
      -  添加光标点击定位功能
v1.3.4:
    date:2020-05-14
    changes:
      -  优化光标控制（增加键盘上下键位置计算）
      -  增加tab键，font-weight等配置接口
      -  优化中文输入的时候输入口显示样式
v1.4.0:
    date:2020-05-16
    changes:
      -  预留lang配置选项，为内置语言提供接口
      -  添加光标选择（删除，tab多行添加等）
v1.4.1:
    date:2020-05-17
    changes:
      -  添加复制、剪切等快捷键
v1.4.2:
    date:2020-05-17
    changes:
      -  修复复制或剪切后光标自动消失问题
v1.5.1:
    date:2020-05-18
    changes:
      -  添加内置着色（支持html+css+javascript）
v1.5.2:
    date:2020-05-19
    changes:
      -  修复初始化选择区域为空导致的错误
      -  添加全选（ctrl + A）功能
v1.5.3:
    date:2020-05-22
    changes:
      -  修复javascript字符串着色错误
      -  添加对css和javascript中边界的着色功能
v1.5.4:
    date:2020-05-25
    changes:
      -  针对css补充着色（完毕）
v1.5.5:
    date:2020-05-26
    changes:
      -  针对javascript补充着色（完毕）
v1.5.6:
    date:2020-05-26
    changes:
      -  补充javascript着色细节
      -  修复ctrl+A功能BUG
      -  修复issue2(https://github.com/yelloxing/Web-Studio-Code/issues/2)
v1.6.0:
    date:2020-06-03
    changes:
      -  添加对标准JSON的支持（着色和格式化开发完毕）
v1.7.0:
    date:2020-06-05
    changes:
      -  HTML格式化（内嵌套的CSS和ES无格式）
v1.7.1:
    date:2020-06-05
    changes:
      -  修复IE和Edge不支持透明色问题
      -  修复IE浏览器对Symbol的支持问题
v1.7.2:
    date:2020-06-06
    changes:
      -  解决错误的HTML模板格式化卡顿问题
v1.7.4:
    date:2020-06-12
    changes:
      -  添加文本内容改变通知钩子`this.updated(callback);`
v1.7.5:
    date:2020-06-28
    changes:
      -  针对编辑器滚动后聚焦位置错误进行了修复
v1.8.0:
    date:2020-07-01
    changes:
      -  使用新的方式改写了着色后页面更新方式，大大提高了效率
      -  第一次引入diff模式来选择性更新页面内容而不是一次性全部重新生成
v1.8.1:
    date:2020-07-01
    changes:
      -  修复在特殊情况下匹配的重复利用行数为负的情况
v1.9.0:
    date:2020-07-03
    changes:
      -  添加输入辅助方法（input配置提供：https://github.com/yelloxing/Web-Studio-Code/issues/9）
      -  对整体的UI进行比较小的优化
v2.0.0:
    date:2020-07-27
    changes:
      -  废弃内置的语言支持，专心提供通用的编辑器

src/index.js

import xhtml from './xhtml';

import isElement from '@yelloxing/core.js/isElement';
import isString from '@yelloxing/core.js/isString';
import isFunction from '@yelloxing/core.js/isFunction';

// 核心方法和工具方法

import { textWidth, bestLeftNum, calcCanvasXY, selectIsNotBlank, toTemplate } from './edit-view/tool';

import { initDom, initView } from './edit-view/init';
import { updateView, updateSelectView, updateCursorPosition, updateCanvasSize, cancelSelect, deleteSelect } from './edit-view/update';
import bindEvent from './edit-view/bind';
import diff from './edit-view/diff';

import filterText from './edit-view/filter';

let wscode = function (options) {

    if (!(this instanceof wscode)) {
        throw new Error('WSCode is a constructor and should be called with the `new` keyword');
    }

    /**
     * 
     * [格式化配置]
     * 
     * 所有的配置校验和默认值设置等都应该在这里进行
     * 经过这里处理以后，后续不需要再进行校验了
     * 因此这里的内容的更改一定要慎重
     * 
     */

    // 编辑器挂载点
    if (isElement(options.el)) {

        // 着色器
        let shader = () => {
            let resultData = [];
            this._contentArray.forEach(text => { resultData.push([{ content: text, color: this._colorText }]); });
            return resultData;
        };

        // 格式化
        let format = textString => textString;

        this._el = options.el;

        // 公共配置
        options.color = options.color || {};
        this._colorBackground = options.color.background || "#d6d6e4"; /*编辑器背景*/
        this._colorText = options.color.text || "#000000"; /*普通文本颜色*/
        this._colorNumber = options.color.number || "#888484"; /*行号颜色*/
        this._colorEdit = options.color.edit || "#eaeaf1"; /*编辑行颜色*/
        this._colorCursor = options.color.cursor || "#ff0000"; /*光标颜色*/
        this._colorSelect = options.color.select || "#6c6cf1"; /*选择背景*/
        this._fontFamily = options["font-family"] || "新宋体"; /*字体*/
        this._fontWeight = options["font-weight"] || 600;/*字重*/
        this._tabSpace = options.tabSpace || 4;/*设置一个tab表示多少个空格*/

        // 文本
        this._contentArray = isString(options.content) ? (this.$$filterText(options.content) + "").split("\n") : [""];

        // 着色方法
        this.$shader = isFunction(options.shader) ? options.shader : shader;

        // 格式化方法
        this.$format = isFunction(options.format) ? options.format : format;

        // 辅助输入
        this.$input = isFunction(options.input) ? options.input : null;

    } else {

        // 挂载点是必须的，一定要有
        throw new Error('options.el is not a element!');
    }

    // 先初始化DOM
    this.$$initDom();

    // 初始化控制变量
    this.__needUpdate = true;
    this.__lineNum = this._contentArray.length - 1;
    this.__leftNum = this._contentArray[this.__lineNum].length;
    this.__cursor1 = this.__cursor2 = { leftNum: 0, lineNum: 0 };
    this.__formatData = this.$$diff(this.$shader(this._contentArray.join('\n')));

    // 初始化视图
    this.$$initView();

    // 更新视图
    this.$$updateView();

    // 绑定操作
    this.$$bindEvent();

    this.__updated__ = () => { };
    // 编辑器管理的文本发生改变后会主动触发callback方法
    this.updated = callback => {
        this.__updated__ = callback;
    };

    // 获取当前编辑器代码
    this.valueOf = () => {
        return this._contentArray.join('\n');
    };

    // 在当前光标位置输入新的内容
    this.input = (content = "", cursor = 0, number = 0) => {

        // 先删除多余的内容

        if (cursor != 0) {

            if (number != 0) {
                this._contentArray[this.__lineNum] =
                    this._contentArray[this.__lineNum].substring(0, this.__leftNum + cursor) +
                    this._contentArray[this.__lineNum].substring(this.__leftNum + cursor + number);
            }

            // 修改光标位置
            this.__leftNum += cursor;

        }

        // 输入以触发更新
        this.__focusDOM.value = content;
        xhtml.trigger(this.__focusDOM, 'input');
        this.__focusDOM.focus();

    };

    // 格式化代码
    this.format = () => {

        // 格式化内容
        this._contentArray = this.$format(this._contentArray.join('\n'), this._tabSpace).split('\n');

        this.__lineNum = this._contentArray.length - 1;
        this.__leftNum = this._contentArray[this.__lineNum].length;

        // 着色
        this.__formatData = this.$$diff(this.$shader(this._contentArray.join('\n')));

        // 更新视图
        this.$$updateView();

        // 更新光标位置
        this.$$initView();

    };

};

// 挂载辅助方法
wscode.prototype.$$textWidth = textWidth;
wscode.prototype.$$bestLeftNum = bestLeftNum;
wscode.prototype.$$calcCanvasXY = calcCanvasXY;
wscode.prototype.$$selectIsNotBlank = selectIsNotBlank;
wscode.prototype.$$filterText = filterText;
wscode.prototype.$$toTemplate = toTemplate;

// 挂载核心方法

wscode.prototype.$$initDom = initDom;
wscode.prototype.$$initView = initView;

wscode.prototype.$$updateView = updateView;
wscode.prototype.$$updateSelectView = updateSelectView;
wscode.prototype.$$updateCursorPosition = updateCursorPosition;
wscode.prototype.$$updateCanvasSize = updateCanvasSize;
wscode.prototype.$$cancelSelect = cancelSelect;
wscode.prototype.$$deleteSelect = deleteSelect;

wscode.prototype.$$bindEvent = bindEvent;

// 性能优化系列方法

wscode.prototype.$$diff = diff;

wscode.author = '心叶（yelloxing@gmail.com）';

if (typeof module === "object" && typeof module.exports === "object") {
    module.exports = wscode;
} else {
    window.WSCode = wscode;
}